### PR TITLE
More MRI MaskGenerators

### LINKED ISSUE #204: AccelerationMaskGenerator mask size
`deepinv.physics.generator.AccelerationMaskGenerator` seems to generate mask tensors of size (B, 2, H, W) whereas the MRI class takes as input sizes of (B, 1, H, W).

### PR DESCRIPTION
**Pull request by Andrewwango**
Add Cartesian mask generators for:
- Uniform random sampling `RandomMaskGenerator` (i.e. existing `deepinv.physics.generator.mri.AccelerationMaskGenerator`)
- Gaussian sampling `GaussianMaskGenerator`
- Uniform (equispaced) sampling `UniformMaskGenerator`
- Base class for enforcing same acceleration factor logic
- Support for k-t sampling in all the above, e.g. with random time-varying mask or sheared mask etc.

Also fix #204 

### Checks to be done before submitting your PR
- [x] `python3 -m pytest tests/` runs successfully.
- [x] `black .` runs successfully.
- [x] `make html` runs successfully (in the `docs/` directory).
- [x] Updated docstrings related to the changes (as applicable).
- [x] Added an entry to the [CHANGELOG.rst](../CHANGELOG.rst).

### CODE REVIEW

**Comment by eseaoutdoorsuk** [comment]
TODO: add Julian's changes from #206

**Comment by Andrewwango** [comment]
TODO: add Julian's changes from #206

File: deepinv/tests/test_generators.py [hunk 1]
```diff
@@ -4,10 +4,10 @@
 import itertools
 
 # Generators to test (make sure they appear in find_generator as well)
+# Note MRI mask generators tests now in test_mri_generators.py
```
Comment (samuro95) [inline deepinv/tests/test_generators.py:7]: we need to remove this line now

Comment (samuro95) [reply deepinv/tests/test_generators.py:7]: the rest looks great, thanks !

**Review by samuro95 — APPROVED** [review]

File: deepinv/physics/generator/mri.py [hunk 2]
```diff
@@ -1,83 +1,244 @@
...
+            self.W // 2 - self.n_center // 2 : self.W // 2 + ceildiv(self.n_center, 2)
+        ] = 0
+
+        # normalise distribution
+        pdf /= np.sum(pdf)
+
+        # select low-frequency lines according to pdf
+        for b in range(mask.shape[0]):
+            for t in range(mask.shape[2]):
+                idx = self.rng.choice(self.W, self.n_lines, replace=False, p=pdf)
```
Comment (tachella) [inline deepinv/physics/generator/mri.py:142]: I have the feeling that this is going to be a bit slow for large batch sizes

Can we use torch.randperm(self.W)[:self.n_lines] ?

See [this post](https://stackoverflow.com/questions/59461811/random-choice-with-pytorch)

File: deepinv/physics/generator/mri.py [hunk 3]
```diff
@@ -1,83 +1,244 @@
...
+
+    The mask is repeated across channels and randomly varies across batch dimension.
+
+    Algorithm taken from Schlemper et al. `A Deep Cascade of Convolutional Neural Networks for Dynamic MR Image Reconstruction <https://github.com/js3611/Deep-MRI-Reconstruction/blob/master/utils/compressed_sensing.py>`_.
+
+    For parameter descriptions see :class:`deepinv.physics.generator.mri.BaseMaskGenerator`
+
+    Example:
+
+    ::
```
Comment (tachella) [inline deepinv/physics/generator/mri.py:174]: same here for doctests

File: deepinv/physics/generator/mri.py [hunk 4]
```diff
@@ -1,83 +1,244 @@
...
+    Supports k-t sampling, where the uniform mask is sheared across time.
 
-        return {"mask": torch.cat([mask.float().unsqueeze(1)] * C, dim=1)}
+    The mask is repeated across channels and randomly varies across batch dimension. Based off fastMRI code https://github.com/facebookresearch/fastMRI
+
+    For parameter descriptions see :class:`deepinv.physics.generator.mri.BaseMaskGenerator`
+
+    Example:
+
+    ::
```
Comment (tachella) [inline deepinv/physics/generator/mri.py:216]: same here for doctests

File: deepinv/physics/generator/mri.py [hunk 5]
```diff
@@ -1,83 +1,244 @@
...
+    """Generator for MRI Cartesian acceleration masks using uniform (equispaced) non-random undersampling.
+
+    Generate a mask of vertical lines for MRI acceleration with fixed sampling in low frequencies (center of k-space) and equispaced undersampling in the high frequencies.
+
+    The number of lines selected with equal spacing are at a proportion that reaches the desired acceleration rate taking into consideration the number of low-freq lines, so that the total number of lines is (N / acceleration).
+
+    Supports k-t sampling, where the uniform mask is sheared across time.
 
-        return {"mask": torch.cat([mask.float().unsqueeze(1)] * C, dim=1)}
+    The mask is repeated across channels and randomly varies across batch dimension. Based off fastMRI code https://github.com/facebookresearch/fastMRI
```
Comment (tachella) [inline deepinv/physics/generator/mri.py:210]: it is a bit confusing that the mask has 'non-random undersampling' but 'randomly varies across batch dimension'?

File: deepinv/physics/generator/mri.py [hunk 6]
```diff
@@ -1,83 +1,244 @@
...
+
+        # determine acceleration rate by adjusting for the number of low frequencies
+        adjusted_accel = (self.acc * (self.n_center - self.W)) / (
+            self.n_center * self.acc - self.W
+        )
+        offset = self.rng.integers(0, round(adjusted_accel))
+
+        for b in range(mask.shape[0]):
+            for t in range(mask.shape[2]):
+                accel_samples = (
```
Comment (tachella) [inline deepinv/physics/generator/mri.py:237]: I'm also worried here about speed

File: deepinv/physics/generator/mri.py [hunk 7]
```diff
@@ -1,83 +1,244 @@
...
-            H // 2 + num_lines_center // 2 + 1,
-            steps=50,
-            dtype=torch.long,
+        if self.T == 0:
+            mask = mask[:, :, 0, :, :]
+
+        return {"mask": mask}
+
+
+class RandomMaskGenerator(BaseMaskGenerator):
```
Comment (tachella) [inline deepinv/physics/generator/mri.py:99]: Should we call this RandomUniformMaskGenerator? 

Or merge UniformMaskGenerator and RandomMaskGenerator into UniformMaskGenerator with a random=True/False option in the init?

**Review by tachella — CHANGES_REQUESTED** [review]
The PR looks very good! 

I have a small concern about speed regarding using np.random.choice in nested for loops - maybe there is a better fix using torch directly?

The doctests are not running at the moment (small change to >>> is needed) 

I'm also wondering whether we can merge (or improve) the naming of the uniform vs non-random uniform classes?

File: deepinv/physics/generator/mri.py (hunk 2 above)
Comment (Andrewwango) [reply deepinv/physics/generator/mri.py:142]: This might be tricky - you would have to use `torch.multinomial(pdf, n_lines, replacement=False)` - but this still won't help with vectorising the loop, which I think is needed as you want the acceleration rate to be same each time-step and each image in batch dimension.

File: deepinv/physics/generator/mri.py (hunk 6 above)
Comment (Andrewwango) [reply deepinv/physics/generator/mri.py:249]: I don't think doing `np.arange`s per batch is a big problem - it should be quite quick?

File: deepinv/physics/generator/mri.py (hunk 2 above)
Comment (tachella) [reply deepinv/physics/generator/mri.py:146]: ok, we can keep it like this and maybe optimise in the future.

File: deepinv/physics/generator/mri.py (hunk 6 above)
Comment (tachella) [reply deepinv/physics/generator/mri.py:249]: ok

**Review by tachella — APPROVED** [review]
